soroban-env-host/src/host.rs

```diff
@@ -804,15 +805,13 @@ impl EnvBase for Host {
         res
     }
 
-    fn symbol_new_from_slice(&self, s: &str) -> Result<SymbolObject, HostError> {
+    fn symbol_new_from_slice(&self, s: &[u8]) -> Result<SymbolObject, HostError> {
         call_env_call_hook!(self, s.len());
         self.charge_budget(ContractCostType::MemCmp, Some(s.len() as u64))?;
-        for ch in s.chars() {
-            SymbolSmall::validate_char(ch)?;
-        }
-        let res = self.add_host_object(ScSymbol(
-            self.metered_slice_to_vec(s.as_bytes())?.try_into()?,
-        ));
+        let res = self.add_host_object(ScSymbol::try_from_bytes(
+            self,
+            self.metered_slice_to_vec(s)?.try_into()?,
+        )?);
```

Comment:
I noticed @graydon's patch he posted [here](https://stellarfoundation.slack.com/archives/C06ATA4RPRP/p1704397276350619) created a common way for all mem host object types to define themselves to be fallibly convertable from bytes. The change in this PR works too, and less abstractions ftw, but @graydon may have a reason for us introducing that abstraction and it may be valuable.